GPU/Directx9/VertexShaderGeneratorDX9.cpp

```diff
@@ -279,6 +284,18 @@ void GenerateVertexShaderDX9(int prim, char *buffer, bool useHWTransform) {
 	}
 	WRITE(p, "};\n");
 
+	if (!gstate.isModeThrough()) {
+		// Apply the projection and viewport to get the Z buffer value, floor to integer, undo the viewport and projection.
+		// Not completely sure this is 100% right under DX9 as the Z range is different...
```

Comment:
I think we have to change the scale here?  For example, if scale and center were 32767.5 (perfectly normal values, btw) then in OpenGL:

z [-1,1] * 32767.5 + 32767.5
=> [0, 65535]

But in Direct3D (this is after the projection multiply):

z [0,1] * 32767.5 + 32767.5
=> [32767.5, 65535]

This will lose an extra bit worth of integer precision.

So what we want is really:

z * (scale * 2) - (65535.0 / 2 + center)

I think?

-[Unknown]